preprocessing/structs.ts

```diff
@@ -1,15 +1,82 @@
+
+   Bins are laid in sequence. If all the offers in a bin have tick `t`, all the offers in the following bin have tick `t+tickSpacing`.
+
+   Bins are identified by their index in the bin sequence, from the first (`MIN_BIN`) to the last (`MAX_BIN`). The number of available bins is larger than the number of available ticks, so some bins will never be used.
+
+   ### Tree structure
+
+   The structure of the tick tree is as follows:
+   <img src="./tick_tree_structure.png" width="190%"></img>
+
+   At the bottom of the tree, leaves contain information about 4 bins: their first and last offer. Offer ids use 32 bits, so leaves use 256 bits.
+
+   When a market order runs, execution starts with the first offer of the smallest nonempty bin and continues from there.
+
+   Once all the offers in the smallest bin have been executed, the next nonempty bin is found. If the leaf of the current bin now only has empty bins, the node above the current leaf contains a bitfield with information about all its children: if the *i*th bit of the node is set, its *i*th child has at least one nonempty bin. Otherwise, its *i*th child only has empty bins.
+
+   To find the next nonempty bin, it may be necessary to keep going up the tree until the root is reached. At each level above, the bitfield rule applies similarly: the *i*th bit of a node is set iff its *i*th child has at least one set bit.
+
+   Once a node with a set bit is found, its rightmost nonempty child is examined, and so on until the next nonempty bin is reached, and the first offer of that bin gets executed.
+
+   ## Caching
+
+   At any time, if there is at least one offer in the offer list, the best offer is the first offer of the bin containing the cheapest offers; and that bin is the best bin. Its parent is the best `level3`, whose parent is the best `level2`, and whose parent is the best `level3` (whose parent is the `root`). 
+
+   The `root`, the best `level1`, the  best`level2`, and the best `level3` are always stored in `local`. The position of the best bin in the best leaf is also stored in `local`. 
+
+   This data is useful for two things:
+   - Read and modify the tree branch of the best offer without additional storage reads and writes (except for modifying the best leaf).
+   - Know the price of the best offer without an additional storage read (it can be computed using the set bit positions in each level, the position of the best bin in the best leaf, and the value of `tickSpacing`).
+
+   The structure of the local config is as follows:
+   <img src="./local_config.png" width="140%"></img>
+
+  This caching means that as the price oscillates within a more restricted range, the fewer additional storage read/writes have to be performed (as most of the information is available in `local`) when there is a market order or an offer insertion/update.
+
+  ## Some numbers
+  Here are some useful numbers, for reference:
+   <img src="./numbers.png" width="70%"></img>
```

Comment:
Nit: Spacing could be improved:

![billede](https://github.com/mangrovedao/mangrove-core/assets/3458856/5da63ea1-f01f-4b66-845e-f5effe13ec3c)

`min_price` could also be interesting to have